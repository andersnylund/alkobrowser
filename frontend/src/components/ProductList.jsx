import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { Loader, Icon, Message } from 'semantic-ui-react';
import styled from 'styled-components';
import produce from 'immer';
import { connect } from 'react-redux';
import { string } from 'prop-types';

import Button from './Button';
import Product from './Product';
import { titleCase } from '../helpers';

const Wrapper = styled.section`
  max-width: 900px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const PRODUCT_QUERY = gql`
  query products(
    $endCursor: String
    $orderBy: ProductOrderByInput
    $where: ProductWhereInput
  ) {
    productsConnection(
      where: $where
      first: 5
      orderBy: $orderBy
      after: $endCursor
    ) {
      pageInfo {
        startCursor
        endCursor
      }
      edges {
        node {
          id
          nimi
          valmistaja
          pullokoko
          hinta
          litrahinta
          tyyppi {
            tyyppi
          }
          luonnehdinta
          pakkaustyyppi
          alkoholiprosentti
          alkoholilitrahinta
        }
      }
    }
  }
`;

export const createWhere = (selectedField, search, selectedCategory) => {
  const where = {
    [`${selectedField}_not`]: null,
    OR: [
      // TODO fix hack
      { nimi_contains: search },
      { nimi_contains: search.toUpperCase() },
      { nimi_contains: search.toLowerCase() },
      { nimi_contains: titleCase(search) },
      { luonnehdinta_contains: search },
      { luonnehdinta_contains: search.toUpperCase() },
      { luonnehdinta_contains: search.toLowerCase() },
      { luonnehdinta_contains: titleCase(search) },
      { tyyppi: { tyyppi_contains: search } },
      { tyyppi: { tyyppi_contains: search.toUpperCase() } },
      { tyyppi: { tyyppi_contains: search.toLowerCase() } },
      { tyyppi: { tyyppi_contains: titleCase(search) } },
      { valmistaja_contains: search },
      { valmistaja_contains: search.toUpperCase() },
      { valmistaja_contains: search.toLowerCase() },
      { valmistaja_contains: titleCase(search) },
    ],
  };

  if (selectedCategory !== 1) {
    where.tyyppi = {
      id: selectedCategory,
    };
  }
  return where;
};

export const ProductList = ({
  selectedField,
  selectedCategory,
  sort,
  search,
}) => {
  const variables = {
    endCursor: null,
    orderBy: `${selectedField}_${sort}`,
    where: createWhere(selectedField, search, selectedCategory),
  };
  return (
    <Wrapper>
      <Query query={PRODUCT_QUERY} variables={variables}>
        {({ data, loading, error, fetchMore }) => {
          if (loading) {
            return <Loader active data-testid="loader" />;
          }
          if (error) {
            return <Message negative>{error.message}</Message>;
          }
          return (
            <>
              {data.productsConnection.edges.map(edge => (
                <Product key={edge.node.id} product={edge.node} />
              ))}
              {data.productsConnection.pageInfo.endCursor && (
                <Button
                  onClick={() => {
                    fetchMore({
                      variables: {
                        endCursor: data.productsConnection.pageInfo.endCursor,
                      },
                      updateQuery: (prev, { fetchMoreResult }) => {
                        if (!fetchMoreResult) {
                          return prev;
                        }
                        const nextState = produce(prev, draft => {
                          // eslint-disable-next-line no-param-reassign
                          draft.productsConnection.pageInfo.endCursor =
                            fetchMoreResult.productsConnection.pageInfo.endCursor;
                          draft.productsConnection.edges.push(
                            ...fetchMoreResult.productsConnection.edges
                          );
                        });

                        return nextState;
                      },
                    });
                  }}
                >
                  <Icon name="plus" />
                  Lisää
                </Button>
              )}
            </>
          );
        }}
      </Query>
    </Wrapper>
  );
};

ProductList.propTypes = {
  selectedField: string.isRequired,
  selectedCategory: string.isRequired,
  sort: string.isRequired,
  search: string.isRequired,
};

const mapStateToProps = state => ({
  selectedField: state.filter.selectedField,
  selectedCategory: state.filter.selectedCategory,
  sort: state.filter.sort,
  search: state.filter.search,
});

export default connect(mapStateToProps)(ProductList);
