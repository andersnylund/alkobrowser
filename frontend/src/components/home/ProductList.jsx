import React, { useEffect } from 'react';
import { Loader, Icon, Message } from 'semantic-ui-react';
import styled from 'styled-components';
import produce from 'immer';
import { connect } from 'react-redux';
import { string } from 'prop-types';
import posed from 'react-pose';

import Button from '../common/Button';
import Product from './Product';
import { getProducts as getProductsAction } from '../../actions/productActions';

const Container = styled.section`
  max-width: 900px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const PosedContainer = posed(Container)({
  enter: { staggerChildren: 10 },
});

const PosedItem = posed.div({
  enter: { x: 0, opacity: 1 },
  exit: { x: 50, opacity: 0 },
});

export const ProductList = ({
  selectedField,
  selectedCategory,
  sort,
  search,
  getProducts,
  isLoading,
  products,
}) => {
  useEffect(() => {
    getProducts();
  }, [getProducts]);

  if (isLoading) {
    return <Loader active />;
  }

  return (
    <PosedContainer>
      {products.map(product => (
        <PosedItem key={product.id}>
          <Product product={product} />
        </PosedItem>
      ))}
    </PosedContainer>
  );
};

ProductList.propTypes = {
  selectedField: string.isRequired,
  selectedCategory: string.isRequired,
  sort: string.isRequired,
  search: string.isRequired,
};

const mapStateToProps = state => ({
  selectedField: state.filter.selectedField,
  selectedCategory: state.filter.selectedCategory,
  sort: state.filter.sort,
  search: state.filter.search,
  isLoading: state.product.isLoading,
  products: state.product.products,
});

const mapDispatchToProps = {
  getProducts: getProductsAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductList);
