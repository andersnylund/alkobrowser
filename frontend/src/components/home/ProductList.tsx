import React, { useEffect, FC } from 'react';
import { Loader, Icon, Message } from 'semantic-ui-react';
import styled from 'styled-components';
import produce from 'immer';
import { connect } from 'react-redux';
import { string } from 'prop-types';
import posed from 'react-pose';

import Button from '../common/Button';
import Product from './Product';
import { getProducts as getProductsAction } from '../../actions/productListActions';
import { AppState } from '../../store';
import { Product as ProductType } from '../../../../shared/types';

const Container = styled.section`
  max-width: 900px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const PosedContainer = posed(Container)({
  enter: { staggerChildren: 10 },
});

const PosedItem = posed.div({
  enter: { x: 0, opacity: 1 },
  exit: { x: 50, opacity: 0 },
});

export interface Props {
  getProducts: () => void;
  isLoading: boolean;
  products?: ProductType[];
}

export const ProductList: FC<Props> = ({ getProducts, isLoading, products }) => {
  useEffect(() => {
    getProducts();
  }, [getProducts]);

  if (isLoading || !products) {
    return <Loader active />;
  }

  // TODO: show error message

  return (
    <PosedContainer>
      {products.map((product: ProductType) => (
        <PosedItem key={product.id}>
          <Product product={product} />
        </PosedItem>
      ))}
    </PosedContainer>
  );
};

const mapStateToProps = (state: AppState) => ({
  isLoading: state.productList.isLoading,
  products: state.productList.products,
});

const mapDispatchToProps = {
  getProducts: getProductsAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductList);
