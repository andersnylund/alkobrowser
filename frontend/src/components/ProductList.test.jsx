import React from 'react';
import { render, cleanup } from '@testing-library/react';
import 'jest-dom/extend-expect';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import { ProductList, PRODUCTLIST_QUERY, createWhere } from './ProductList';

describe('<ProductList />', () => {
  beforeEach(cleanup);

  it('should render without an error', () => {
    render(
      <MockedProvider>
        <ProductList
          selectedField="hinta"
          selectedCategory="1"
          search=""
          sort="ASC"
        />
      </MockedProvider>
    );
  });

  it('should show a loading indicator', () => {
    const { getByTestId } = render(
      <MockedProvider>
        <ProductList
          selectedField="hinta"
          selectedCategory="1"
          search=""
          sort="ASC"
        />
      </MockedProvider>
    );
    const loader = getByTestId('loader');
    expect(loader).toHaveAttribute('class', 'ui active loader');
  });

  it('should show an error message', async () => {
    const errorMock = {
      request: {
        query: PRODUCTLIST_QUERY,
        variables: {
          endCursor: null,
          orderBy: 'hinta_ASC',
          where: createWhere('hinta', '', '1'),
        },
      },
      error: new Error('Testing error'),
    };

    const { getByText } = render(
      <MockedProvider mocks={[errorMock]} addTypename={false}>
        <ProductList
          selectedField="hinta"
          selectedCategory="1"
          search=""
          sort="ASC"
        />
      </MockedProvider>
    );

    await wait(0);

    expect(getByText('Network error: Testing error')).toHaveTextContent(
      'Network error: Testing error'
    );
  });

  it('should not show any products', async () => {
    const productMock = {
      request: {
        query: PRODUCTLIST_QUERY,
        variables: {
          endCursor: null,
          orderBy: 'hinta_ASC',
          where: createWhere('hinta', '', '1'),
        },
      },
      result: {
        data: {
          productsConnection: {
            edges: [],
            pageInfo: {
              endCursor: null,
              startCursor: null,
            },
          },
        },
      },
    };

    const { queryByTestId } = render(
      <MockedProvider mocks={[productMock]} addTypename={false}>
        <ProductList
          selectedField="hinta"
          selectedCategory="1"
          search=""
          sort="ASC"
        />
      </MockedProvider>
    );

    await wait(0);

    expect(queryByTestId('product')).toBeNull();
  });

  it('should show a list of products', async () => {
    const productMock = {
      request: {
        query: PRODUCTLIST_QUERY,
        variables: {
          endCursor: null,
          orderBy: 'hinta_ASC',
          where: createWhere('hinta', '', '1'),
        },
      },
      result: {
        data: {
          productsConnection: {
            edges: [
              {
                node: {
                  id: '1',
                  nimi: 'nimi',
                  valmistaja: 'valmistaja',
                  pullokoko: 0.56,
                  hinta: 123,
                  litrahinta: 'litrahinta',
                  tyyppi: {
                    id: 'id',
                    tyyppi: 'tyyppi',
                  },
                  luonnehdinta: 'luonnehdinta',
                  pakkaustyyppi: 'pakkaustyyppi',
                  alkoholiprosentti: 3.1,
                  alkoholilitrahinta: 123.123,
                },
              },
            ],
            pageInfo: {
              endCursor: null,
              startCursor: null,
            },
          },
        },
      },
    };

    const { container } = render(
      <MockedProvider mocks={[productMock]} addTypename={false}>
        <ProductList
          selectedField="hinta"
          selectedCategory="1"
          search=""
          sort="ASC"
        />
      </MockedProvider>
    );

    await wait(0);

    expect(container).toMatchSnapshot();
  });
});
