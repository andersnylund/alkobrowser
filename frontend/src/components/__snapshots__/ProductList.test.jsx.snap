// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductList /> should show a list of products 1`] = `
<div>
  <section
    class="sc-bZQynM dqIhmd"
  >
    <a
      href="/products/1"
    >
      <div
        class="sc-bwzfXH liXxLE"
        style="transform: none;"
      >
        <div
          class="sc-htpNat gkAJbA"
        >
          <img
            alt="nimi"
            src="https://images.alko.fi/images/cs_srgb,f_auto,t_medium/cdn/1/nimi.jpg"
          />
        </div>
        <div
          class="sc-bxivhb goNXdw"
        >
          <div
            class="sc-ifAKCX kQXLog"
          >
            <h3>
              nimi
            </h3>
            <p>
              luonnehdinta
            </p>
          </div>
          <div
            class="sc-EHOje cCuemI"
          >
            <p>
              tyyppi
            </p>
            <table>
              <tbody>
                <tr>
                  <td>
                    Valmistaja
                  </td>
                  <td>
                    valmistaja
                  </td>
                </tr>
                <tr>
                  <td>
                    Alkoholiprosentti
                  </td>
                  <td>
                    3.1 %
                  </td>
                </tr>
                <tr>
                  <td>
                    Hinta
                  </td>
                  <td>
                    123 €
                  </td>
                </tr>
                <tr>
                  <td>
                    Alkoholin litrahinta
                  </td>
                  <td>
                    123.123 €/l
                  </td>
                </tr>
                <tr>
                  <td>
                    Litrahinta
                  </td>
                  <td>
                    litrahinta €/l
                  </td>
                </tr>
                <tr>
                  <td>
                    Pakkaustyyppi
                  </td>
                  <td>
                    pakkaustyyppi
                  </td>
                </tr>
                <tr>
                  <td>
                    Pullokoko
                  </td>
                  <td>
                    0.56 litraa
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </a>
  </section>
</div>
`;
